### DISCUSSION TITLE
Using CLI parser inside a library

### DISCUSSION THREAD

**Discussion by marcmellanox** [CLIUtils/CLI11#753] [Q&A]
Hi,

Is it possible to ue CLI library inside a library.
I would like to use it to parse environment variables.
A library doesn't have a main entry, but some init function.would it be possible to do something like 
CLI::App mylib{"My library"};
mylib.add_option("--my_option", opt)->envname("MY_OPTION");

CLI11_PARSE(mylib): <- with no argc and argv

Thanks

**Comment by marcmellanox** [comment]
I made something like this and it works fine.

int mylib_init(){

 CLI::App lib{"Library"};
        lib.add_option("my_option", opt)
                ->envname("MY_OPTION")
                ->check(CLI::Range(0,100));


        const char *argv[]= {"My library"};


  try   {
          lib.parse(1, (argv));
        } catch(const CLI::ParseError &e) {
        return ERROR;
        }
}
But still waiting for suggestion if someone already did it differently

**Comment by phlptp** [comment]
We do this all the time inside [HELICS](https://github.com/GMLC-TDC/HELICS/search?q=CLI11).  
The first layer usually gets passed the argc, argv from the command line, but subsequent layers will also use the single string parser which does the splitting as well and a vectorized version which takes a `std::vector<string>` instead of the `argv` structure. 

CLI11 is also used for parsing config files.

In your case if you were just using it to capture environment variables, you can probably just call

```
try {
lib.parse("");
} catch(const CLI::ParseError &e) {
return ERROR;
}
}
```
